worker/wallet.js

```diff
@@ -1,36 +1,165 @@
+            return resolve()
+          }
+        } catch (error) {
+          logEventError('hodl_invoice_updated', error)
+          reject(error)
+        }
+      })
+      sub.on('error', reject)
+    })
+  } catch (error) {
+    console.error(error)
+  }
+  sub?.removeAllListeners()
+}
 
-// TODO this should all be done via websockets
-export async function checkInvoice ({ data: { hash, isHeldSet }, boss, models, lnd }) {
+export async function finalizeHodlInvoice ({ data: { hash }, models, lnd }) {
+  let inv
+  try {
+    inv = await getInvoice({ id: hash, lnd })
+  } catch (err) {
+    console.log(err, hash)
+    return
+  }
+  if (!inv.is_confirmed) await cancelHodlInvoice({ id: hash, lnd })
+}
+
+async function subscribeToWithdrawals (args) {
+  const { lnd } = args
+
+  // https://www.npmjs.com/package/ln-service#subscribetopayments
+  const sub = subscribeToPayments({ lnd })
+  sub.on('confirmed', async (payment) => {
+    logEvent('confirmed', payment)
+    try {
+      await checkWithdrawal({ data: { hash: payment.id }, ...args })
+    } catch (error) {
+      logEventError('confirmed', error)
+    }
+  })
+  sub.on('failed', async (payment) => {
+    logEvent('failed', payment)
+    try {
+      await checkWithdrawal({ data: { hash: payment.id }, ...args })
+    } catch (error) {
+      logEventError('failed', error)
+    }
+  })
+  // ignore payment attempts
+  sub.on('paying', (attempt) => {})
+  sub.on('error', console.error)
+
+  // check pending withdrawals since they might have been paid while worker was down.
+  await checkPendingWithdrawals(args)
+}
+
+async function checkPendingWithdrawals (args) {
+  const { models } = args
+  const pendingWithdrawals = await models.withdrawl.findMany({ where: { status: null } })
+  for (const w of pendingWithdrawals) {
+    await checkWithdrawal({ data: { id: w.id, hash: w.hash }, ...args })
+  }
+}
+
+async function checkPendingDeposits (args) {
+  const { models } = args
+  const pendingDeposits = await models.invoice.findMany({ where: { confirmedAt: null, cancelled: false } })
+  for (const d of pendingDeposits) {
+    await checkInvoice({ data: { id: d.id, hash: d.hash }, ...args })
+  }
+}
+
+async function checkInvoice ({ data: { hash }, boss, models, lnd }) {
   let inv
   try {
     inv = await getInvoice({ id: hash, lnd })
   } catch (err) {
     console.log(err, hash)
-    // on lnd related errors, we manually retry so we don't exponentially backoff
-    await boss.send('checkInvoice', { hash }, walletOptions)
     return
   }
   console.log(inv)
 
-  // check if invoice still exists since HODL invoices get deleted after usage
+  // check if invoice exists since it might just have been created by LND and wasn't inserted into the database yet
+  // but that is not a problem since this function will be called again with the update
+  // FIXME: there might be a race condition here if the invoice gets paid before the invoice was inserted into the db.
```

Comment:
I think the probability of users paying invoices before it got inserted in the db is so low, we can basically ignore this case.

However, the chance that `checkInvoice` gets called before the invoice is inserted is pretty high. However, it doesn't really matter since `checkInvoice` on invoices that were just created wouldn't do anything anyway.

I could check if I can skip such `checkInvoices` call though. Maybe via a check if the invoice was created within the last few milliseconds or something. But not sure if that's worth it. Will make the code more messy again without a lot of benefit.

Reply:
How would they pay an invoice we haven't shown them? Delivering the invoice to the client requires it being stored in the db.

Reply:
> How would they pay an invoice we haven't shown them? Delivering the invoice to the client requires it being stored in the db.

Ah, good point, haha